File: A2Z_striver/patterns/pattern_9/pattern_9.cpp
#include <iostream>

using namespace std;

// Header Comment Block
// Problem: Print a Solid Diamond (2N rows).
// Example Input: n = 3
// Example Output:
//   *
//  ***
// *****
// *****
//  ***
//   *
// Goal: O(N^2) Time | O(1) Space

class Solution
{
public:
    // ~ Approach 1: Stacked Pyramids (Your Solution)
    // Analysis: Time O(N^2) | Space O(1)
    // Pros: Very readable, easy to debug.
    void pattern9_stacked(int n)
    {
        // 1. Top Half
        for (int i = 0; i < n; i++)
        {
            for (int j = 0; j < n - i - 1; j++)
                cout << " ";
            for (int j = 0; j < 2 * i + 1; j++)
                cout << "*";
            cout << "\n";
        }
        // 2. Bottom Half
        for (int i = 0; i < n; i++)
        {
            for (int j = 0; j < i; j++)
                cout << " ";
            for (int j = 0; j < 2 * (n - i) - 1; j++)
                cout << "*";
            cout << "\n";
        }
    }

    // ~ Approach 2: Single Loop (Index Mirroring)
    // Analysis: Time O(N^2) | Space O(1)
    // Pros: Follows DRY (Don't Repeat Yourself) principle.
    // Cons: Logic is slightly more complex to derive on the spot.
    void pattern9_single_loop(int n)
    {
        // - We run one loop for the total 2*n rows
        for (int i = 0; i < 2 * n; i++)
        {

            // ! Key Logic: "The Mirror"
            // - If we are in the top half (i < n), use 'i' as is.
            // - If we are in the bottom half (i >= n), flip 'i' to count backwards.
            //   Formula: (2*n - i - 1) converts row indices [n, 2n-1] -> [n-1, 0]
            int row = i;
            if (i >= n)
            {
                row = 2 * n - i - 1;
            }

            // 2. Single Inner Loop
            // Limit: We need to print spaces + stars.
            // Spaces count: n - row - 1
            // Stars count:  2 * row + 1
            // Total columns needed: (n - row - 1) + (2 * row + 1) = n + row
            for (int j = 0; j < n + row + 1; j++)
            { // +1 to cover the last star? No, < n+row covers indices 0 to n+row-1.
                // Wait, let's re-verify the math:
                // Row 0 (n=3): Needs 2 spaces, 1 star. Total 3 chars. Indices 0, 1, 2. Limit should be 'n + row'.

                // 3. Conditional Logic
                // Spaces are on the left side, count is (n - row - 1).
                // So if index j < count, it's a space.
                if (j < n - row - 1)
                {
                    cout << " ";
                }
                else
                {
                    cout << "*";
                }
            }
            cout << "\n";
        }
    }
};

int main()
{
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    Solution sol;
    int n = 3;

    cout << "--- Approach 1: Stacked ---\n";
    sol.pattern9_stacked(n);

    cout << "\n--- Approach 2: Single Loop (Mirrored) ---\n";
    sol.pattern9_single_loop(n);

    return 0;
}

File: A2Z_striver/1 Basics/1.2 Patterns/pattern_9/pattern_9.py
import sys

class Solution:
    """
    Problem: Print a Solid Diamond (2N rows).
    Example Input: n = 3
    Example Output:
      *
     ***
    *****
    *****
     ***
      *
    Goal: O(N^2) Time.
    """

    # // ~ Approach 1: Stacked Pyramids (Standard/Readable)
    # // Analysis: Time O(N^2) | Space O(1)
    def pattern9_stacked(self, n):
        # 1. Top Half (Upright)
        for i in range(n):
            # Spaces
            for _ in range(n - i - 1):
                print(" ", end="")
            # Stars
            for _ in range(2 * i + 1):
                print("*", end="")
            print()

        # 2. Bottom Half (Inverted)
        for i in range(n):
            # Spaces
            for _ in range(i):
                print(" ", end="")
            # Stars
            for _ in range(2 * (n - i) - 1):
                print("*", end="")
            print()

    # // ~ Approach 2: Single Loop (Index Mirroring)
    # // Analysis: Time O(N^2) | Space O(1)
    def pattern9_single_loop(self, n):
        # Loop runs 2*n times (0 to 2n-1)
        for i in range(2 * n):
            
            # ! Key Logic: "The Mirror"
            # If i >= n, we flip the index to count backwards
            effective_row = i
            if i >= n:
                effective_row = 2 * n - i - 1
            
            # Print Spaces: n - effective_row - 1
            for _ in range(n - effective_row - 1):
                print(" ", end="")
            
            # Print Stars: 2 * effective_row + 1
            for _ in range(2 * effective_row + 1):
                print("*", end="")
            
            print()

# Driver Code
if __name__ == "__main__":
    solver = Solution()
    n = 3
    
    print(f"--- Approach 1: Stacked (n={n}) ---")
    solver.pattern9_stacked(n)
    
    print(f"\n--- Approach 2: Single Loop (n={n}) ---")
    solver.pattern9_single_loop(n)